### PR TITLE
178 save sample info in firebase store

### PR DESCRIPTION
**Pull request by rkunal**
Create/Update Sample collection in firestore whenever sample's metadata.json is stored in cloud firebase

Update: 
1. SampleCollection entity should always be dependent on Sample. I dont forsee it being used in isolation.
2. I have avoided any abstractions for Firestore related operations for now. As, I do not know how it would be used in future. Any architecture changes related to that are welcome and will be implemented as part of this PR.

### CODE REVIEW

**Comment by rkunal** [comment]
1. Currently using metadata.json timestamp format eg, 2022-06-22T21:10:19+0530. Instead of firestore's timestamp data type. Need to change
2. sampleID field is added in-order to fetch and update the collection

File: app/src/main/java/net/aiscope/gdd_app/network/FirestoreUtil.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package net.aiscope.gdd_app.network
+import com.google.firebase.firestore.FirebaseFirestore
+
+class FirestoreUtil {
```
Comment (simtop) [inline app/src/main/java/net/aiscope/gdd_app/network/FirestoreUtil.kt:4]: Hello, why do you create a new instance for Firebase Firestore? You could use FirebaseRemoteStorage to store the sample, that has an instance of the Firebase FireStore

File: app/src/main/java/net/aiscope/gdd_app/model/SampleModels.kt [hunk 2]
```diff
@@ -31,6 +33,17 @@ data class Sample(
     fun nextImageName(): String = "${id}_image_${captures.completedCaptureCount()}"
 
     fun nextMaskName(): String = "${id}_mask_${captures.completedCaptureCount()}"
+
+    fun buildSampleCollection() : SampleCollection {
```
Comment (simtop) [inline app/src/main/java/net/aiscope/gdd_app/model/SampleModels.kt:37]: you can one line it doing fun buildSampleCollection() = SampleCollection(
            createdOn = Timestamp(createdOn.time),
            uploadedBy = FirestoreUtil.FirestoreUtil.firestore?.
            collection("microscopist")?.document(microscopist),
            location = id,
            numberOfImages = captures.completedCaptureCount()
        )

File: app/src/main/java/net/aiscope/gdd_app/ui/sample_completion/SampleCompletionViewModel.kt [hunk 3]
```diff
@@ -146,6 +146,7 @@ class SampleCompletionViewModel @Inject constructor(
                     microscopistRepository.load().copy(hasSubmitSampleFirstTime = true)
                 )
             }
+            remoteStorage.storeSampleCollection(storedSample)
```
Comment (simtop) [inline app/src/main/java/net/aiscope/gdd_app/ui/sample_completion/SampleCompletionViewModel.kt:149]: Is this needed? Having it we will end up saving 2 samples each time, I think modifying the code just on top of this to accept the new Model will solve the issue.

**Review by simtop — CHANGES_REQUESTED** [review]

File: app/src/main/java/net/aiscope/gdd_app/ui/sample_completion/SampleCompletionViewModel.kt (hunk 3 above)
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/sample_completion/SampleCompletionViewModel.kt:149]: They end up in different locations though. This new 'storedsample' goes to https://console.firebase.google.com/u/0/project/aiscope-labelling-app-test/firestore/data/~2Fsamples~2F3a03bbd6-4047-4094-8c4a-3e6fa0863898

As opposed to https://console.firebase.google.com/u/0/project/aiscope-labelling-app-test/storage/aiscope-labelling-app-test.appspot.com/files/~2F3a03bbd6-4047-4094-8c4a-3e6fa0863898

File: app/src/main/java/net/aiscope/gdd_app/network/FirestoreUtil.kt (hunk 1 above)
Comment (rkunal) [reply app/src/main/java/net/aiscope/gdd_app/network/FirestoreUtil.kt:4]: I could not find firebaseFirestore instance in FirebaseRemoteStorage. FirebaseModule exists, but I am not sure how to use it. Unless I move all DB operations to SampleRepositorySharedPreference ?

PS: To understand things better, I have simplified the code, moved all DB operations to one SampleViewModel and commented everything else. Where should this code be best moved to ?

Comment (rkunal) [reply app/src/main/java/net/aiscope/gdd_app/network/FirestoreUtil.kt]: @simtop  I have added a new repository to store sample in firestore. Currently it does not implement  samplerepository interface.  feel it's too early for that.  Cleaned up a bit. Let me know.

**Review by simtop — APPROVED** [review]

**Review by jbors — APPROVED** [review]

**Pull request merged**